cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1822,6 +1823,20 @@ private RunRecordDetail scanRunByReference(Collection<Field<?>> keys)
     }
   }
 
+  /**
+   * Deletes all completed run records with a start time before {@code timeUpperBound}, throwing
+   * {@code IOException} if the delete operation fails.
+   */
+  public void deleteRunsStartedBefore(Instant timeUpperBound) throws IOException {
+    long invertedStartTime = getInvertedTsKeyPart(timeUpperBound.getEpochSecond());
+    ImmutableList<Field<?>> fields = ImmutableList.of(
+            Fields.stringField(StoreDefinition.AppMetadataStore.RUN_STATUS, TYPE_RUN_RECORD_COMPLETED),
+            Fields.longField(StoreDefinition.AppMetadataStore.RUN_START_TIME, invertedStartTime));
+
+    // Delete all records *after* the inverted start time (equivalent to *before* the provided start time)
+    getRunRecordsTable().deleteAll(Range.from(fields, Range.Bound.INCLUSIVE));
```

Comment:
`Range.from(fields, Range.Bound.INCLUSIVE)` is creating a range that has the lowerBound but no upperBound, which could cause the bug that returns satisfied rows which meet condition `RUN_STATUS >= TYPE_RUN_RECORD_COMPLETED && RUN_START_TIME >= invertedStartTime`. 

Suggest to use the existing method `createRunRecordScanRange` that has both lowerBound and upperBound, something like this:
```
ImmutableList<Field<?>> keyPrefix = ImmutableList.of(
        Fields.stringField(StoreDefinition.AppMetadataStore.RUN_STATUS, TYPE_RUN_RECORD_COMPLETED));

// Delete all records *after* the inverted start time (equivalent to *before* the provided start time)
Range range = createRunRecordScanRange(keyPrefix, 0L, timeUpperBound.getEpochSecond());
getRunRecordsTable().deleteAll(range);
```

Reply:
Makes sense, updated.